### PR TITLE
Add multi-dataset fine-tuning via MultiTrainer

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Adds multi-dataset fine-tuning: passing a list or tuple of datasets to `model.train()` fine-tunes a single base model on each dataset in series via a new `MultiTrainer`, returning a `{dataset: metrics}` mapping and saving a cross-dataset results plot. This generalizes the old `RF100Benchmark` helper into a mechanism that works for any dataset collection (RF100, custom sets, etc.).

```python
from ultralytics import YOLO

model = YOLO("yolo26n.pt")
results = model.train(data=["coco8.yaml", "african-wildlife.yaml"], epochs=10)
results["coco8.yaml"]["fitness"]  # per-dataset metrics; also saves runs/<task>/multitrain_results.png
```

## Changes

- **`MultiTrainer`** (`engine/trainer.py`): fine-tunes the same base weights on each dataset independently (a fresh copy seeds every run), accumulates per-dataset metrics, and plots a cross-dataset bar chart. A failed dataset is logged and skipped so a sweep still completes.
- **`Model.train()`** dispatches to `MultiTrainer` when `data` is a list/tuple; the single-dataset path is unchanged.
- **DDP metrics fix**: `Model.train()` now recovers training metrics from the saved checkpoint when validation runs in a DDP subprocess, fixing the long-standing `None` return under multi-GPU (previously a `# TODO`) for both single- and multi-dataset training.
- **`plot_multitrain_results`** (`utils/plotting.py`): cross-dataset bar chart of the per-dataset metric with the mean line.
- **RF100**: the `RF100Benchmark` utility is removed; its Roboflow download flow is migrated into `docs/en/datasets/detect/roboflow-100.md` as a self-contained script that uses the new `model.train(data=[...])` system (no more stdout log-scraping).

## Testing

- New `test_train_multi` (gated on `ONLINE`, skipped on edge devices).
- Verified the single-dataset `train`/`val` paths are unchanged (still return their task metrics objects).
- `ruff` clean; reference docs regenerated.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 This PR adds built-in multi-dataset training support, making it easier to fine-tune one YOLO model across many datasets in a single `model.train()` call.

### 📊 Key Changes
- Added a new `MultiTrainer` class that automatically runs when `data` is passed as a list or tuple to `model.train()`.
- Updated `YOLO.train()` to return a `{dataset: metrics}` dictionary for multi-dataset runs, instead of only single-run metrics.
- Grouped all multi-dataset outputs under one `runs/<task>/multitrain/` directory, including:
  - per-dataset subfolders
  - `multitrain_results.json`
  - `multitrain_results.png` 📈
- Added automatic summary plotting with a new `plot_multitrain_results()` utility.
- Improved distributed training behavior so training metrics are now recovered from checkpoints when needed, ensuring results are returned more reliably on DDP/GPU setups ✅
- Added `normalize_platform_uri()` so pasted Ultralytics Platform web URLs are automatically converted to `ul://` URIs.
- Updated dataset loading utilities so Platform dataset/model URLs work directly in training inputs.
- Removed the older `RF100Benchmark` utility and replaced its docs with a simpler workflow based on standard `YOLO.train()` usage.
- Refreshed documentation and tests to cover:
  - multi-dataset training
  - direct Platform URL support
  - new result-saving behavior

### 🎯 Purpose & Impact
- Makes large benchmark or dataset-sweep workflows much simpler by replacing custom scripts with one training call ⚡
- Helps users evaluate a base model across many datasets consistently, with each run starting from the same initial weights.
- Improves usability for Roboflow 100-style experiments by providing built-in aggregation, visualization, and machine-readable results.
- Reduces friction when using the [Ultralytics Platform](https://platform.ultralytics.com) by allowing users to paste dataset or model URLs directly 🌐
- Improves reliability for multi-GPU training by ensuring metrics are returned even in distributed setups.
- Overall, this change makes cross-dataset experimentation more automated, organized, and accessible for both researchers and everyday users 🎯

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review; an engineer will assist soon. Please review the checklist below to support a smooth merge ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -1084,3 +1084,85 @@ def build_optimizer(self, model, name="auto", lr=0.001, momentum=0.9, decay=1e-5
...
+            LOGGER.info(f"\n{colorstr('blue', 'bold', f'MultiTrainer {i + 1}/{len(datasets)}:')} fine-tuning on {data}")
+            try:
+                overrides = {**self.args, "data": data, "name": Path(str(data)).stem, "resume": False, "session": None}
+                trainer = self.trainer(overrides=overrides, _callbacks=self.callbacks)
+                trainer.model = trainer.get_model(weights=deepcopy(self.model), cfg=self.model.yaml)
+                trainer.train()
+                self.trainers.append(trainer)
+                # Read metrics from the saved checkpoint so results survive the DDP training subprocess (multi-GPU)
+                ckpt = trainer.best if trainer.best.exists() else trainer.last
+                self.metrics[data] = torch_load(ckpt)["train_metrics"] if ckpt.exists() else None
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: `self.metrics` is keyed by the dataset value itself, so repeated entries silently overwrite earlier runs. The new test already uses `['coco8.yaml', 'coco8.yaml']`, which means both trainers run but only the last result survives in the returned dict. This makes the advertised `{dataset: metrics}` API incomplete for duplicate datasets and loses run order.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The multi-dataset training flow looks well integrated overall, but there are a couple of correctness issues around result aggregation/visualization collisions, plus an unintended breaking API removal of `RF100Benchmark` that should be handled more gracefully.

💬 Posted 3 inline comments

File: ultralytics/utils/benchmarks.py [hunk 2]
```diff
@@ -269,147 +264,6 @@ def benchmark(
     return df_display
 
 
-class RF100Benchmark:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/benchmarks.py:272]: ⚠️ **HIGH**: Removing `RF100Benchmark` outright is a breaking public API change: it was documented and importable from `ultralytics.utils.benchmarks`, so existing user code will now fail with `ImportError`. If the new multi-dataset flow is the replacement, please keep a deprecated compatibility shim or alias for at least one release.

File: ultralytics/engine/trainer.py [hunk 3]
```diff
@@ -1084,3 +1084,85 @@ def build_optimizer(self, model, name="auto", lr=0.001, momentum=0.9, decay=1e-5
...
+                self.plot_results()
+        return self.metrics
+
+    def plot_results(self):
+        """Save a cross-dataset bar chart of the per-dataset metric with the mean across all datasets."""
+        from ultralytics.cfg import TASK2METRIC
+        from ultralytics.utils.plotting import plot_multitrain_results
+
+        key = TASK2METRIC.get(self.args.get("task"))
+        scores = {Path(str(d)).stem: float(m.get(key, m.get("fitness", 0.0))) for d, m in self.metrics.items() if m}
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: 💡 **MEDIUM**: Using `Path(str(d)).stem` for the plot labels can merge distinct datasets that share the same filename, such as `foo/data.yaml` and `bar/data.yaml`, into a single bar. That means `multitrain_results.png` can silently drop runs even when `self.metrics` kept them separate. The plot should use a unique per-run label instead of the stem alone.

File: ultralytics/engine/trainer.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/engine/trainer.py]: Fixed in a99b7be: results are now keyed by the unique run-directory name (e.g. `coco8`, `coco8-2`) instead of the dataset value, so repeated datasets no longer collapse — each run appears separately in the returned dict, `multitrain_results.json`, and the plot.

File: ultralytics/engine/trainer.py (hunk 3 above)
Comment (glenn-jocher) [reply ultralytics/engine/trainer.py]: Fixed in a99b7be: the plot now uses the same unique per-run keys (`coco8`, `coco8-2`, etc.) rather than `Path(stem)`, so distinct datasets sharing a filename get distinct bars.

File: ultralytics/utils/benchmarks.py (hunk 2 above)
Comment (glenn-jocher) [reply ultralytics/utils/benchmarks.py:272]: This removal is intentional: `RF100Benchmark` (a thin helper whose only consumer was its own docs example) is replaced by the general `model.train(data=[...])` multi-dataset flow, and its Roboflow download logic is migrated into the Roboflow 100 docs page so the documented workflow still runs. Given it served only that one example, we're removing it directly rather than carrying a shim.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Huge thanks to @glenn-jocher for this excellent addition.

“Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example of that idea in action: multi-dataset training, cleaner result aggregation, better DDP reliability, and smoother [Ultralytics Platform](https://platform.ultralytics.com) dataset/model URL support all come together to make YOLO workflows much more powerful and accessible. A meaningful step forward for large-scale benchmarking and everyday training alike.
